balloon/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+//! Pure Rust implementation of the [Balloon] password hashing function as
+//! specified in [this paper](https://eprint.iacr.org/2016/027.pdf).
+//!
+//! # Usage (simple with default params)
+//!
+//! Note: this example requires the `rand_core` crate with the `std` feature
+//! enabled for `rand_core::OsRng` (embedded platforms can substitute their
+//! own RNG)
+//!
+//! Add the following to your crate's `Cargo.toml` to import it:
+//!
+//! ```toml
+//! [dependencies]
+//! balloon-hash = "0.1"
+//! rand_core = { version = "0.6", features = ["std"] }
+//! sha2 = "0.9"
+//! ```
+//!
+//! The following example demonstrates the high-level password hashing API:
+//!
+//! ```
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # #[cfg(all(feature = "password-hash", feature = "std"))]
+//! # {
+//! use balloon_hash::{
+//!     password_hash::{
+//!         rand_core::OsRng,
+//!         PasswordHash, PasswordHasher, PasswordVerifier, SaltString
+//!     },
+//!     Balloon
+//! };
+//! use sha2::Sha256;
+//!
+//! let password = b"hunter42"; // Bad password; don't actually use!
+//! let salt = SaltString::generate(&mut OsRng);
+//!
+//! // Balloon with default params
+//! let balloon = Balloon::<Sha256>::default();
+//!
+//! // Hash password to PHC string ($balloon$v=1$...)
+//! let password_hash = balloon.hash_password(password, &salt)?.to_string();
+//!
+//! // Verify password against PHC string
+//! let parsed_hash = PasswordHash::new(&password_hash)?;
+//! assert!(balloon.verify_password(password, &parsed_hash).is_ok());
+//! # }
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! [Balloon]: https://en.wikipedia.org/wiki/Balloon_hashing
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![doc(
+    html_logo_url = "https://raw.githubusercontent.com/RustCrypto/media/8f1a9894/logo.svg",
+    html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/media/8f1a9894/logo.svg",
+    html_root_url = "https://docs.rs/balloon-hash/0.1.0"
+)]
+#![warn(rust_2018_idioms, missing_docs)]
+
+#[cfg(feature = "alloc")]
+extern crate alloc;
+
+mod error;
+mod params;
+
+pub use crate::{
+    error::{Error, Result},
+    params::Params,
+};
+use core::convert::{TryFrom, TryInto};
+use core::marker::PhantomData;
+use digest::generic_array::GenericArray;
+use digest::Digest;
+#[cfg(feature = "password-hash")]
+#[cfg_attr(docsrs, doc(cfg(feature = "password-hash")))]
+pub use password_hash::{self, PasswordHash, PasswordHasher, PasswordVerifier};
+#[cfg(all(feature = "alloc", feature = "password-hash"))]
+use password_hash::{Decimal, Ident, ParamsString, Salt};
+
+/// Balloon context.
+///
+/// This is the primary type of this crate's API, and contains the following:
+///
+/// - Default set of [`Params`] to be used
+/// - (Optional) Secret key a.k.a. "pepper" to be used
+#[derive(Clone, Default)]
+pub struct Balloon<'key, D: Digest> {
```

Comment:
I decided to store the hash algorithm to use in `Balloon` because I can't pass it into `PasswordHasher` otherwise.

I could also move it to `Params`, but I was unsure if I should store the hash algorithm in the PHC string and I don't know how to store it in the PHC string anyway, where do I get an ID or name for every hash algorithm from?

Another idea would be to make `PasswordHasher` always use SHA-256, or some other algorithm.

Reply:
A thought: the key size is fixed to the size of a `block_t`, which is the hash digest size. So instead of making `Balloon` depend on a `'key` lifetime, you can just make the secret a `GenericArray<u8, D::OutputSize>` or something

Reply:
re the hasher: parameterizing by `D` and using a `PhantomData` is a pretty standard pattern

Reply:
> What precisely would be the purpose of a "pepper" in this scheme? Honest question because I've never seen it used.

I thought the reason it was used in Argon2 was because it is considered good practice. See [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#peppering) for example.

> re the hasher: parameterizing by `D` and using a `PhantomData` is a pretty standard pattern

Yes, I'm familiar with it, I'm just uncertain because which hash algorithm is used is quiet important, considering `Params` isn't storing it, it's being left out of the PHC string too. In any case, as noted in Zulip, we might wanna remove the whole `PasswordHasher` anyway.